test/MinimalDelegation.execute.t.sol

```diff
@@ -445,76 +406,86 @@ contract MinimalDelegationExecuteTest is TokenHandler, DelegationHandler, Execut
+        Call[] memory calls = CallUtils.init();
         calls = calls.push(buildTransferCall(address(tokenA), address(receiver), 1e18)); // Transfer 1 tokenA
         calls = calls.push(buildTransferCall(address(tokenB), address(receiver), 1e18)); // Transfer 1 tokenB
 
-        // Get the current nonce components for key 0
-        uint256 nonceKey = 0;
-        (uint256 nonce,) = _buildNextValidNonce(nonceKey);
+        bytes memory executionData = abi.encode(calls, _signAndPack(hash(calls, DEFAULT_NONCE), signerTestKey, DEFAULT_NONCE, ROOT_KEY_HASH));
 
-        // Create hash of the calls + nonce and sign it
-        SignedCalls memory signedCalls = SignedCalls({calls: calls, nonce: nonce});
-        bytes32 hashToSign = signerAccount.hashTypedData(signedCalls.hash());
-        (uint8 v, bytes32 r, bytes32 s) = vm.sign(signerPrivateKey, hashToSign);
-        bytes memory signature = abi.encodePacked(r, s, v);
+        signerAccount.execute(BATCHED_CALL_SUPPORTS_OPDATA, executionData);
+        vm.snapshotGasLastCall("execute_BATCHED_CALL_SUPPORTS_OPDATA_twoCalls");
+    }
 
-        // Pack the execution data:
-        // 1. Encode the nonce and signature into opData
-        bytes memory opData = abi.encode(nonce, signature);
-        // 2. Encode the calls and opData together
-        bytes memory executionData = abi.encode(calls, opData);
+    /**
+     * Edge case tests
+     */
+    function test_execute_batch_emptyCalls_succeeds() public {
+        Call[] memory calls = CallUtils.init();
+        vm.prank(address(signerAccount));
+        signerAccount.execute(BATCHED_CALL, abi.encode(calls));
+    }
+
+    function test_execute_batch_emptyCalls_revertsWhenUnauthorized() public {
+        Call[] memory calls = CallUtils.init();
+        vm.expectRevert(IERC7821.Unauthorized.selector);
+        signerAccount.execute(BATCHED_CALL, abi.encode(calls));
+    }
+
+    /**
+     * Self call tests
+     */
+    function test_execute_register_update_asRoot_succeeds() public {
+        TestKey memory newKey = TestKeyManager.initDefault(KeyType.Secp256k1);
+        Call[] memory calls = CallUtils.init();
+        calls = CallUtils.addRegister(calls, newKey);
+        calls = CallUtils.addUpdate(calls, newKey.toKeyHash(), Settings.wrap(0));
+
+        vm.prank(address(signerAccount));
+        signerAccount.execute(BATCHED_CALL, abi.encode(calls));
+    }
+
+    function test_execute_register_update_asNonRoot_reverts() public {
+        TestKey memory newKey = TestKeyManager.initDefault(KeyType.Secp256k1);
+        Call[] memory calls = CallUtils.init();
+        calls = CallUtils.addRegister(calls, newKey);
+        calls = CallUtils.addUpdate(calls, newKey.toKeyHash(), Settings.wrap(0));
+
+        vm.expectRevert(IERC7821.Unauthorized.selector);
+        signerAccount.execute(BATCHED_CALL, abi.encode(calls));
+    }
+
+    function test_execute_register_update_withRootSignature_succeeds() public {
```

Comment:
initDefault isn't the root key though ?

Reply:
good catch

Reply:
oh its because the root key is calling the functions to register the newKey and update it